### PR TITLE
Fixed backgrounds should behave like 'scroll' inside transformed elements

### PR DESCRIPTION
**Pull request by vitorroriz**
#### eba490ba3cc4ca38142884808f25545819227472
<pre>
Fixed backgrounds should behave like &apos;scroll&apos; inside transformed elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=15679">https://bugs.webkit.org/show_bug.cgi?id=15679</a>
rdar://problem/99926239

Reviewed by Antti Koivisto.

* LayoutTests/TestExpectations:
Enabling related tests that would fail before.

* LayoutTests/compositing/backgrounds/fixed-background-on-descendant.html:
* LayoutTests/compositing/backgrounds/fixed-backgrounds.html:
These tests need fixed-background but and they make use of transforms. After this patch,
the tests would no longer render background as fixed, so we are updating them to
avoid transforms but still trigger the creation of composited layer.

* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::calculateBackgroundImageGeometry):

According to spec (<a href="https://drafts.csswg.org/css-transforms/)">https://drafts.csswg.org/css-transforms/)</a>, if an element or any of its ancestor elements
have a transform, its background-attachment property should be treated as it has &apos;scroll&apos; value instead of &apos;fixed&apos;.

Canonical link: <a href="https://commits.webkit.org/255055@main">https://commits.webkit.org/255055@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/e9fe5826b2994cbe37bffbfc6fbb82eb3ab097ac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91066 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/54 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21600 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/100509 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/159555 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95072 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/62 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29125 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83422 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97190 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96722 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/52 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77838 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27026 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82003 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/81657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70060 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35215 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15679 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33012 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16677 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3513 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36794 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/39633 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38721 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35761 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/rendering/RenderBoxModelObject.cpp [hunk 1]
```diff
@@ -1218,9 +1218,13 @@ BackgroundImageGeometry RenderBoxModelObject::calculateBackgroundImageGeometry(c
     LayoutSize positioningAreaSize;
     // Determine the background positioning area and set destination rect to the background painting area.
     // Destination rect will be adjusted later if the background is non-repeating.
-    // FIXME: transforms spec says that fixed backgrounds behave like scroll inside transforms. https://bugs.webkit.org/show_bug.cgi?id=15679
     LayoutRect destinationRect(borderBoxRect);
-    bool fixedAttachment = fillLayer.attachment() == FillAttachment::FixedBackground;
```
Comment (mattwoodrow) [inline Source/WebCore/rendering/RenderBoxModelObject.cpp]: This code has moved into `BackgroundPainter::calculateBackgroundImageGeometry`, would you be able to rebase against latest trunk?

I'd also expect `LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/transform-fixed-bg-002/4/5/6/7.html` to pass with this change, do you need to update the `TestExpectations` file?

Other than that, looks good to me!

Comment (vitorroriz) [reply Source/WebCore/rendering/RenderBoxModelObject.cpp]: rebased

File: Source/WebCore/rendering/BackgroundPainter.cpp [hunk 2]
```diff
@@ -512,9 +512,11 @@ BackgroundImageGeometry BackgroundPainter::calculateBackgroundImageGeometry(cons
     LayoutSize positioningAreaSize;
     // Determine the background positioning area and set destination rect to the background painting area.
     // Destination rect will be adjusted later if the background is non-repeating.
-    // FIXME: transforms spec says that fixed backgrounds behave like scroll inside transforms. https://bugs.webkit.org/show_bug.cgi?id=15679
+    auto enclosing = renderer.enclosingLayer();
+    bool isTransformed = renderer.hasTransform() || (enclosing ? enclosing->hasTransformedAncestor() : false);
```
Comment (nt1m) [inline Source/WebCore/rendering/BackgroundPainter.cpp]: Can we rename `enclosing` to `enclosingLayer`?

also the condition can be:
```
    bool isTransformed = renderer.hasTransform() || (enclosing && enclosing->hasTransformedAncestor());
```

File: Source/WebCore/rendering/BackgroundPainter.cpp [hunk 3]
```diff
@@ -512,9 +512,11 @@ BackgroundImageGeometry BackgroundPainter::calculateBackgroundImageGeometry(cons
...
+    bool fixedAttachment = fillLayer.attachment() == FillAttachment::FixedBackground && !isTransformed && fillLayer.image();
```
Comment (nt1m) [inline Source/WebCore/rendering/BackgroundPainter.cpp]: (just a question): why are we now checking for `fillLayer.image()` ?

**Comment by mattwoodrow** [comment]
Can you please enable the `LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/transform-fixed-bg-002/4/5/6/7.html` tests (by removing their entries from `LayoutTests/TestExpectations`).

I believe these should all now pass with your changes, but unfortunately the EWS runs don't include tests that are expected to fail, so they don't show up as unexpected passes. Running `run-webkit-tests imported/w3c/web-platform-tests/css/css-transforms` locally should show them as unexpected passes.

File: Source/WebCore/rendering/BackgroundPainter.cpp (hunk 3 above)
Comment (vitorroriz) [reply Source/WebCore/rendering/BackgroundPainter.cpp]: Actually I don't think we do, removing it.

**Review by anttijk — APPROVED** [review]

**Pull request merged**